### PR TITLE
Fixes multiple database parsing in ChemicalCompositionAction (#25216)

### LINKED ISSUE #25216: Thermochimica parsing error with multiple instances of ChemicalCompositionAction
## Bug Description
`ChemicalCompositionAction` parses thermodynamic data files using Thermochimica every time it is executed. If an instance of the action is already created and the object is called again, for example when called as a sub-app by the Stochastic Tools module, it leads to a Fortran runtime error where already allocated variables are getting reallocated. 

## Steps to Reproduce
Run `MoRu.i` as a sub-app with GPR surrogate trainer.

## Impact
Causes unnecessary termination of simulations.

## Solution
I'll add a static member variable to track if a database has already been parsed and skip the parsing step when true.

### PR DESCRIPTION
**Pull request by parikshitbajpai**
Fixes broken Valgrind test introduced in #24553 https://github.com/idaholab/moose/pull/24553#issuecomment-1678322953

Also closes #25216

### CODE REVIEW

**Comment by loganharbour** [comment]
> Also closes #52216

Need to fix this reference

**Comment by loganharbour** [comment]
In the future, please keep fixes and new capability separate.

**Comment by parikshitbajpai** [comment]
> In the future, please keep fixes and new capability separate.

I can refactor it and push just the fix. I had a working branch and added the fix there. Not a good choice.

File: modules/chemical_reactions/src/actions/ChemicalCompositionAction.C [hunk 1]
```diff
@@ -102,21 +105,39 @@ ChemicalCompositionAction::ChemicalCompositionAction(const InputParameters & par
...
     if (thermo_file.length() > 1024)
       paramError("thermofile",
-                 "Path exceeds thermochimica's maximal permissible length of 1024 with ",
+                 "Path exceeds Thermochimica's maximal permissible length of 1024 with ",
                  thermo_file.length(),
                  " characters: ",
                  thermo_file);
 
-    Thermochimica::setThermoFilename(thermo_file);
+    if (!(_database_parsed))
```
Comment (lindsayad) [inline modules/chemical_reactions/src/actions/ChemicalCompositionAction.C:116]:
```suggestion
    if (!_database_parsed)
```

File: modules/chemical_reactions/src/actions/ChemicalCompositionAction.C [hunk 2]
```diff
@@ -102,21 +105,39 @@ ChemicalCompositionAction::ChemicalCompositionAction(const InputParameters & par
...
 
-    // Read in thermodynamics model, only once
-    Thermochimica::parseThermoFile();
+      // Read in thermodynamics model, only once
+      Thermochimica::parseThermoFile();
 
-    int idbg = Thermochimica::checkInfoThermo();
-    if (idbg != 0)
-      paramError("thermofile", "Thermochimica data file cannot be parsed. ", idbg);
+      int idbg = Thermochimica::checkInfoThermo();
```
Comment (lindsayad) [inline modules/chemical_reactions/src/actions/ChemicalCompositionAction.C:123]:
```suggestion
      const auto idbg = Thermochimica::checkInfoThermo();
```

File: modules/chemical_reactions/src/actions/ChemicalCompositionAction.C [hunk 3]
```diff
@@ -102,21 +105,39 @@ ChemicalCompositionAction::ChemicalCompositionAction(const InputParameters & par
...
+    }
+    else if (_database_parsed && thermo_file != _database_file)
+      paramError("thermofile",
+                 "Thermodynamic database ",
+                 _database_file,
+                 " already parsed. Cannot parse database ",
+                 thermo_file);
+    else
+    {
+    }
```
Comment (lindsayad) [inline modules/chemical_reactions/src/actions/ChemicalCompositionAction.C:140]:
```suggestion
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
This is small enough and the fix is important; let's just make the requested changes and get this in this time around

**Comment by lindsayad** [comment]
Can you squash the last two commits? Sorry for all the requests

**Comment by parikshitbajpai** [comment]
> Can you squash the last two commits? Sorry for all the requests

Yeah just a sec. doing it

**Review by loganharbour — APPROVED** [review]

**Comment by loganharbour** [comment]
Ah.. sorry didn't see Alex's req. Just talked to @dschwen about how the disgusting statics are appropriate

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
> The following file(s) are changed:
> 
> conda/libmesh/meta.yaml

dunno about that CIVET

**Comment by loganharbour** [comment]
> > The following file(s) are changed:
> > conda/libmesh/meta.yaml
> 
> dunno about that CIVET

It is a little hosed, but it's fine

**Pull request merged**